### ISSUE TITLE
[C0/C1] Hugging Face learner tooling boundary and portable ML artifact stack

### ISSUE THREAD

**Issue by chrismaghuhn** [chrismaghuhn/argentum-engine#137]
# C0/C1 — Hugging Face learner tooling boundary and portable ML artifact stack

## Status

**PLANNING / TOOLING ARCHITECTURE — NOT IMPLEMENTATION AUTHORIZATION**

Current accepted baseline:

```text
origin/main=2a030fa8aa6eb86a1b468f1c7a9ec7f5a10cda89
COMMANDER_ENVIRONMENT_V1_COMPLETE=YES
B0_FINAL_ACCEPTANCE_PASS=YES
B1_FINAL_ACCEPTANCE_PASS=YES
B2_FINAL_ACCEPTANCE_PASS=YES
DATA_TRUSTED=YES
CURRENT_PHASE=C0
C0_AUTHORIZED=YES
C1_AUTHORIZED=NO
TRAINING_AUTHORIZED=NO
```

Related trackers:

- #119 — post-B2 storage / reader / scaling characterization
- #124 — C0+ model/data/evaluation/RL roadmap
- #101 — future portable actor/learner compute

This issue owns only the **replaceable learner/tooling boundary** around Hugging Face and adjacent Python ML infrastructure.

It does **not** own gameplay, trusted-data admission, model architecture selection, training authorization, or large-corpus generation.

---

# Core principle

Hugging Face tooling may own **physical ML execution and transport**, but Argentum remains the semantic authority.

```text
Argentum
  Rules / Gym / Replay
  PlayerObservationV1
  CompleteLegalDomain
  TrajectoryV1
  DatasetManifest / A7 validation
        |
        | trusted canonical source data
        v
Learner adapter / derived view
        |
        +--> HF Datasets / Arrow where justified
        |
        +--> PyTorch learner
                |
                +--> Accelerate where justified
                +--> Trackio for observability
                +--> Safetensors for physical weights
                |
                v
       ArgentumCheckpointManifest

Optional transport:
  huggingface_hub
```

Required invariant:

```text
THIRD_PARTY_TOOLING
!=
ARGENTUM_TRUST_AUTHORITY
```

---

# Argentum remains authoritative for

Hugging Face or any other ML framework must never redefine:

```text
dataset membership / trust
trajectory validity
observation/privacy semantics
complete legal-domain semantics
chosen-in-domain validity
semantic decision identity
replay/determinism
factual episode closure
checkpoint semantic identity
frozen gameplay evaluation
Magic rules / legality
```

A third-party loader/cache may reject malformed physical input, but it must never repair, infer, truncate, relabel or silently reinterpret accepted Argentum source data.

---

# Candidate tooling

## 1. Safetensors — preferred early C1 physical weight container

Provisional direction:

```text
C1_LOCAL_WEIGHTS_FORMAT=safetensors
```

Role:

- physical tensor serialization;
- safe model-weight loading;
- portable local/Kaggle/cloud artifact.

Hard boundary:

```text
weights.safetensors
!=
semantic checkpoint identity
```

The canonical checkpoint must be wrapped by an Argentum-owned versioned manifest, conceptually:

```text
ArgentumCheckpointManifestV1
  checkpoint identity
  model architecture identity
  model config digest
  observation/candidate encoder identity
  sequence contract identity
  source dataset manifest identity
  split identity
  training config digest
  source commit
  parent checkpoint where applicable
  RNG identities
  runtime/framework provenance
  weights content digest
```

Do not treat filename, filesystem path, timestamp, `latest`, `best`, Kaggle run ID, or Hub URL as semantic checkpoint identity.

---

## 2. Trackio — preferred early C1 experiment observability candidate

Provisional role:

```text
metrics / run observability only
```

Candidate metrics:

```text
train loss
validation loss
top-1 chosen-candidate agreement
per-decision-family loss / accuracy
candidate-count distributions
samples/sec
batches/sec
CPU/GPU utilization
memory
checkpoint identity
dataset manifest identity
trajectory schema identity
```

Trackio is **not**:

```text
promotion authority
dataset trust authority
checkpoint identity authority
frozen evaluation authority
```

Experiment logs may reference Argentum semantic identities but do not replace them.

---

## 3. Hugging Face Datasets / Arrow — conditional derived learner view

This is explicitly dependent on #119 measurements.

Potential shape:

```text
accepted immutable TrajectoryV1
→ strict Argentum A7 reader
→ deterministic learner-view materialization
→ Arrow / HF Dataset
→ batching / learner
```

Required classification:

```text
HF_DATASET=DERIVED_VIEW
TRAJECTORY_V1=AUTHORITATIVE_SOURCE
```

A derived Arrow/Parquet/HF Dataset must bind at minimum:

```text
source DatasetManifest identity
source trajectory schema identity
learner-view schema identity
materializer/config identity
content digest
```

It must not:

- decide which episodes are trusted;
- reconstruct missing legal candidates;
- truncate candidates for fixed-width convenience;
- add hidden state;
- change decision chronology;
- randomize rows before deterministic split semantics are applied;
- overwrite canonical TrajectoryV1 source data.

Adoption decision should use #119 evidence for:

```text
canonical bytes/decision
reader throughput
memory use
compression ratio
batch I/O needs
```

Possible outcomes:

```text
HF_DATASETS_ADOPT=YES
HF_DATASETS_ADOPT=NO
HF_DATASETS_ADOPT=DEFER
```

Do not pre-commit before measurement.

---

## 4. Accelerate — candidate for C1 scaled execution

Do not require Accelerate for the first tiny local learner unless it materially simplifies the implementation.

Preferred staging:

```text
C1 local smoke
  PyTorch
  + Trackio
  + Safetensors

then, if local learner contract passes:

C1 scaled
  same PyTorch learner
  + Accelerate
  + Trackio
  + Safetensors
```

Potential roles:

- device placement;
- mixed precision;
- gradient accumulation;
- local GPU -> Kaggle GPU portability;
- later multi-GPU execution if justified.

Accelerate must not change:

```text
model-facing sample semantics
candidate masking semantics
sequence/reset semantics
checkpoint semantic identity
frozen evaluation semantics
```

If different backends produce materially different decisions under a contract that claims deterministic inference, that is an inference-contract issue, not acceptable provider variation.

---

## 5. huggingface_hub — optional artifact transport

Potential later use:

```text
checkpoint transport
derived learner-dataset transport
local <-> Kaggle exchange
research/public artifact hosting
```

Transport is never authority.

Required pattern:

```text
Argentum semantic identity
+ content digest
→ upload / transport
→ download
→ verify
→ consume
```

Do not use mutable aliases such as `latest` as reproducibility identity.

A Hub repository/revision/path must not replace Argentum DatasetManifest or CheckpointManifest identities.

Public hosting decisions remain blocked on measured #119 storage characteristics and explicit release policy.

---

# Tooling deliberately NOT adopted by default

Do not introduce these merely because they are part of the HF ecosystem:

```text
Transformers Trainer
Transformers model architectures
TRL
PEFT / LoRA
Optimum / custom kernels
HF Evaluate as gameplay promotion authority
```

Argentum's expected first model is a structured variable-size entity/candidate scorer, not an LLM fine-tuning workload.

Any future adoption requires an explicit measured need.

---

# Initial C1 physical-stack hypothesis

This issue records the following **provisional** stack for later validation:

```text
C1 local:
  PyTorch
  + Trackio
  + Safetensors

C1 learner data adapter:
  Argentum A7 Reader
  → deterministic derived view
  → HF Datasets / Arrow ONLY IF #119 JUSTIFIES IT

C1 scaled:
  PyTorch
  + Accelerate
  + Trackio
  + Safetensors

artifact transport:
  huggingface_hub OPTIONAL
```

This is a tooling hypothesis, not training authorization.

---

# Interaction with C0 contracts

This issue must consume rather than redefine the contracts frozen by #124/C0:

```text
model-facing sample contract
observation encoding contract
candidate scoring contract
structured decision representation
sequence/reset contract
split contract
evaluation contract
checkpoint identity contract
deterministic inference contract
value/reward boundary
teacher provenance
```

Physical tooling must adapt to those contracts.

The contracts must not be weakened to fit a library API.

---

# Interaction with #119

#119 decides physical-data questions before large-scale adoption.

In particular this issue should consume:

```text
canonical bytes/decision
compressed bytes/decision
storage component breakdown
A7 read throughput
reader memory profile
trusted generation throughput
1M / 10M decision projections
```

Potential decision:

```text
TrajectoryV1 canonical source
→ compressed archive for transport
→ Arrow/HF derived learner cache for training
```

is valid only if measured useful.

Do not redesign TrajectoryV1 merely because an ML library prefers another physical layout.

---

# Future implementation gates

Before implementing the local C1 learner tooling stack require:

```text
C0_MODEL_FACING_SAMPLE_CONTRACT=PASS
C0_CANDIDATE_SCORING_CONTRACT=PASS
C0_SPLIT_CONTRACT=PASS
C0_EVALUATION_CONTRACT=PASS
C0_CHECKPOINT_IDENTITY_CONTRACT=PASS
C1_AUTHORIZED=YES
```

Before adopting HF Datasets/Arrow as a persistent learner cache require additionally:

```text
POST_B2_CHARACTERIZATION=PASS
DERIVED_VIEW_NEED_MEASURED=YES
DERIVED_VIEW_SCHEMA_DEFINED=YES
SOURCE_MANIFEST_BINDING=PASS
NO_CANDIDATE_TRUNCATION=PASS
NO_PRIVACY_DRIFT=PASS
```

Before scaling with Accelerate/Kaggle require:

```text
LOCAL_MINI_MODEL_PIPELINE_PASS=YES
CHECKPOINT_ROUNDTRIP=PASS
TRAIN_INFERENCE_CONTRACT_MATCH=PASS
PROVENANCE_RECORDED=PASS
```

---

# Small learner tooling smoke — future

When separately authorized, the first tooling smoke should remain deliberately tiny:

```text
~1k–10k trusted decisions
```

Purpose:

```text
PIPELINE_VALIDATION
!=
GAMEPLAY_STRENGTH
```

Prove:

```text
TrajectoryV1 reader works
model-facing adapter works
variable-size candidate batching works
no legal candidate is truncated
chosen target exists in supplied domain
finite loss/gradients
tiny overfit works
Safetensors round-trip works
ArgentumCheckpointManifest binds weights exactly
Trackio records the run without becoming authority
deterministic inference contract holds
```

No gameplay-strength claim follows from this smoke.

---

# Security / privacy boundary

Third-party artifact and telemetry tooling must not receive data beyond the accepted learner-facing information set unless separately authorized.

Do not upload or log:

```text
raw GameState
opponent hidden hand/library
private credentials
auth tokens
provider secrets
routing nonces as gameplay features
local private paths as semantic identity
```

Public/external artifact publication requires a separate explicit review.

Local tooling may remain offline/private during early C1 work.

---

# Non-goals

This issue does NOT authorize:

- C1 training;
- Behavior Cloning;
- RL;
- self-play;
- large dataset generation;
- model architecture implementation;
- TrajectoryV1 redesign;
- Arrow/Parquet migration of canonical source data;
- uploading datasets or checkpoints publicly;
- Hugging Face becoming gameplay/data trust authority;
- Transformers/TRL/PEFT adoption by default.

---

# Decision record

Current provisional direction:

```text
SAFETENSORS=PREFERRED_C1_WEIGHT_CONTAINER
TRACKIO=PREFERRED_C1_OBSERVABILITY_CANDIDATE
HF_DATASETS_ARROW=CONDITIONAL_ON_#119
ACCELERATE=DEFER_UNTIL_LOCAL_C1_PIPELINE_PASS
HUGGINGFACE_HUB=OPTIONAL_TRANSPORT_ONLY
TRANSFORMERS_TRAINER=NOT_ADOPTED
TRL=NOT_ADOPTED
PEFT=NOT_ADOPTED

ARGENTUM_DATA_TRUST_AUTHORITY=UNCHANGED
ARGENTUM_CHECKPOINT_AUTHORITY=UNCHANGED
TRAINING_AUTHORIZED=NO
```

Revisit concrete package versions and implementation details only when the corresponding C1 slice is authorized.

**Comment by chrismaghuhn** [comment]
## Learner-tooling status sync — 2026-09-15

This tracker is no longer purely hypothetical: the early local C1 tooling boundary has now been exercised in accepted/near-accepted C1 work.

```text
CURRENT_MAIN=47788a55462ec11080b382e13d78383a506069ac
CURRENT_PHASE=C1

C1_04_LOCAL_LEARNER_TOOLING_FINAL_ACCEPTANCE_PASS=YES
PYTORCH_ADOPTED_FOR_PHYSICAL_LEARNER_EXECUTION=YES
SAFETENSORS_ADOPTED_FOR_PHYSICAL_TENSOR_WEIGHTS=YES
TRACKIO_BOUNDARY_IMPLEMENTED=YES
ARGENTUM_CHECKPOINT_MANIFEST_REMAINS_SEMANTIC_AUTHORITY=YES
```

C1_06 has now used the physical stack for a bounded real-CUDA candidate-scorer smoke:

```text
TORCH_VERSION=2.14.0+cu130
CUDA_SMOKE_EXECUTED=YES
SAFETENSORS_ROUNDTRIP=PASS
DETERMINISTIC_POST_RELOAD_INFERENCE=PASS
C1_06_FINAL_ACCEPTANCE_PASS=NO_PENDING_REMEDIATION_AND_REVIEW
```

The tooling boundary remains unchanged:

```text
THIRD_PARTY_TOOLING != ARGENTUM_TRUST_AUTHORITY
```

Still deferred / conditional:

```text
HF_DATASETS_ARROW=CONDITIONAL_ON_MEASURED_READER_NEED
ACCELERATE=NOT_YET_ADOPTED
HUGGINGFACE_HUB=OPTIONAL_TRANSPORT_ONLY
TRANSFORMERS_TRAINER=NOT_ADOPTED
TRL=NOT_ADOPTED
PEFT=NOT_ADOPTED
```

#188 now has fresh measured relevance because C1_06 spent far more wall-clock in trusted source validation/join/preparation than in the actual bounded GPU optimizer loop. Any Arrow/HF Dataset or other training-optimized cache remains a deterministic derived acceleration artifact only; canonical TrajectoryV1 / accepted derived authority is unchanged.

Do not introduce Accelerate merely for the current single-GPU local smoke. Revisit it when the same learner is moved to a genuinely scaled Kaggle/cloud learner workload or when mixed precision / accumulation / device abstraction has a measured need.

No public upload or Hub dependency is authorized by this update.

**Comment by chrismaghuhn** [comment]
### Tracker disposition

Keep this issue **OPEN**. It now tracks the remaining optional/conditional tooling decisions after the accepted early local stack rather than an unstarted C0 hypothesis.

No additional implementation is authorized by this status-sync comment.
